plugins/catalog-backend/src/service/response/write.ts

```diff
@@ -83,40 +85,59 @@ export async function writeEntitiesResponse(options: {
     const prefix = first ? '[' : ',';
     first = false;
 
-    if (await writeResponseData(res, prefix + entity)) {
+    if ((await writer(prefix + entity)) === 'closed') {
       return;
     }
   }
   res.end(`${first ? '[' : ''}]${trailing}`);
 }
 
 /**
- * Writes a data to the response and waits if the response buffer needs draining.
+ * Creates a data writer that writes to the response and waits if the response
+ * buffer needs draining.
  *
  * @internal
- * @returns true if the response was closed while waiting for the buffer to drain
+ * @returns A writer function. If a write attempt returns 'closed', the
+ * connection has become closed prematurely and the caller should stop trying to
+ * write.
  */
-export async function writeResponseData(res: Response, data: string | Buffer) {
-  const ok = res.write(data, 'utf8');
-  if (!ok) {
+export function createResponseDataWriter(
+  res: Response,
+): (data: string | Buffer) => Promise<'ok' | 'closed'> {
+  // See https://github.com/backstage/backstage/issues/30659
+  //
+  // This code goes to some lengths to just add listeners once at the top,
+  // instead of on every need to drain. Hence it is more complex that seems to
+  // be necessary, just to avoid listener leaks.
+
+  let drainPromise: DeferredPromise<'ok'> | undefined;
+
+  const closePromise = new Promise<'closed'>(resolve => {
+    function onClose() {
+      res.off('drain', onDrain);
+      res.off('close', onClose);
+      res.off('finish', onClose);
+      resolve('closed');
+    }
+    function onDrain() {
+      drainPromise?.resolve('ok');
+      drainPromise = undefined;
+    }
+    res.on('drain', onDrain);
+    res.on('close', onClose);
+    res.on('finish', onClose);
+  });
+
+  return async data => {
+    if (res.write(data, 'utf8')) {
```

Comment:
What do you think about adding a guard before this that makes sure that `drainPromise` is `undefined`? Just to avoid any potential race

Reply:
I could instead change `drainPromise = createDeferred();` to be wrapped in `if (!drainPromise)`. That way, it'll always be safe and listening for whatever the next drain event is for any reason. Right? :)